libs/storage/include/storage/have_seen_recently_cache.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <deque>
+#include <unordered_set>
+
+namespace fetch {
+namespace storage {
+
+/**
+ * Not thread safe.
+ *
+ * This class provides the functionality that you can add elements to it,
+ * they will be held in the order they were added up to a fixed size,
+ * and you can query whether they are still there (have been recently seen).
+ *
+ * Due to the implementation behaviour, adding the same item when it is
+ * already Seen may result in it returning it as not been seen sooner than expected.
+ *
+ */
+template <typename T>
+class HaveSeenRecentlyCache
+{
+public:
+  explicit HaveSeenRecentlyCache(uint64_t size)
+    : size_{size}
+  {}
+
+  bool Seen(T const &item)
+  {
+    return seen_set_.find(item) != seen_set_.end();
+  }
+
+  void Add(T const &item)
+  {
+    seen_set_.insert(item);
+    seen_deque_.push_front(item);
+
+    while (seen_deque_.size() > size_)
```

Comment:
As a side note, should it even be a `while`? In a single thread, can this condition hold for longer than a single iteration?

Reply:
No, but it is good defensive programming practice to do this. For example if you added a setter for the cache size (probably someone will want to do this), it could hold for more than a single iteration immediately after changing the cache size.